### PR TITLE
feat(mocks): provide access to call options in client tests

### LINKED ISSUE #10030: Customers can verify call options
Our APIs look like:
```cc
T Client::Foo(FooRequest const& request, Options options);
T Connection::Foo(FooReqeust const& request);
```

Customers can verify they are giving the right `request` to our library. They cannot verify that they are giving the right `options` to our library.

```cc
void MakeClientCall() {
  // This function can be unit tested by customers
  auto options = MakeCallOptions();

  // But they can never catch the difference between:
  client.Foo(req, options);
  // and...
  client.Foo(req); // oops I forgot to include the options.
}
```

I think the solution is to add:

```cc
namespace google::cloud::mocks {
/**
 * Verify per-call options
 * 
 * @code
 * TEST(Foo, CallOptions) {
 *   auto mock = std::make_shared<MockConnection>();
 *   EXPECT_CALL(*mock, Foo).WillOnce([] {
 *         auto options = google::cloud::mocks::GetCallOptions();
 *         EXPECT_THAT(options, ...);
 *       });
 *   auto client = Client(mock);
 *   DoThing(client);
 * }
 * @endcode
 */
Options GetCallOptions() {
  return internal::CurrentOptions();
}
}
```

### PR DESCRIPTION
**Pull request by dbolduc**
Fixes #10030 

If we think this should be more publicized, say the words and I will add it to our [mocking exemplar](https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/bigquery/samples/mock_bigquery_read.cc).

Also, while we are here, put the other mock test in an anonymous namespace.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11050)

### CODE REVIEW

File: google/cloud/mocks/current_options.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * This would be used to verify configuration options from within a
+ * MockConnection. It provides a way for applications to test the difference
+ * between `client.Foo(request, options)` and `client.Foo(request)`.
+ *
+ * @code
+ * TEST(Foo, CallOptions) {
+ *   auto mock = std::make_shared<MockConnection>();
+ *   EXPECT_CALL(*mock, Foo).WillOnce([] {
+ *         auto const& options = google::cloud::mocks::CurrentOptions();
```
Comment (devbww) [inline google/cloud/mocks/current_options.h:36]: If `CurrentOptions()` is going to return by value, then I wouldn't suggest capturing that value by `const&`.

Comment (dbolduc) [reply google/cloud/mocks/current_options.h:36]: changed to `Options const& mocks::CurrentOptions()`

File: google/cloud/mocks/current_options.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ *         EXPECT_THAT(options, ...);
+ *       });
+ *   auto client = Client(mock);
+ *   DoThing(client);
```
Comment (devbww) [inline google/cloud/mocks/current_options.h:40]: Opt: Perhaps `client->Foo()` would be clearer?

**Review by devbww — APPROVED** [review]

File: google/cloud/mocks/current_options.h (hunk 2 above)
Comment (dbolduc) [reply google/cloud/mocks/current_options.h:40]: It's not quite the same thing. If a customer's test called Foo, they would just verify the options outside of the mock.

```cc
client.Foo(options);
EXPECT_THAT(options, ...);
```

We are helping them unit test their `DoThing` method which could look like:

```cc
void DoThing(Client c) {
  auto options = ...;
  c.Foo(options); // `c.Foo();` would be wrong, but how can I detect the difference?
}
```

That being said, I agree that the link between `DoThing` and `Foo` is unclear / imaginary. I will `s/DoThing/CallFoo/` to make the link clearer.

Comment (dbolduc) [reply google/cloud/mocks/current_options.h]: decided on: `s/DoThing/MyFunctionThatCallsFoo/`

Comment (devbww) [reply google/cloud/mocks/current_options.h]: SGTM.  Thanks.

**Pull request merged**